src/Symfony/Bridge/Doctrine/Validator/DoctrineLoader.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bridge\Doctrine\Validator;
+
+use Doctrine\Common\Persistence\Mapping\MappingException;
+use Doctrine\ORM\Mapping\ClassMetadataFactory;
+use Doctrine\ORM\Mapping\ClassMetadataInfo;
+use Doctrine\ORM\Mapping\MappingException as OrmMappingException;
+use Symfony\Bridge\Doctrine\Validator\Constraints\UniqueEntity;
+use Symfony\Component\Validator\Constraints\Length;
+use Symfony\Component\Validator\Mapping\AutoMappingTrait;
+use Symfony\Component\Validator\Mapping\ClassMetadata;
+use Symfony\Component\Validator\Mapping\Loader\LoaderInterface;
+
+/**
+ * Guesses and loads the appropriate constraints using Doctrine's metadata.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class DoctrineLoader implements LoaderInterface
+{
+    use AutoMappingTrait;
+
+    private $classMetadataFactory;
+    private $namespaces;
+
+    public function __construct(ClassMetadataFactory $classMetadataFactory, array $namespaces = null)
```

Comment:
I see a drawback with injecting the ClassMetadataFactory directly: it is instantiated by the EntityManager itself, and gets a reference to the EM instantiating it. But with our usage of a proxy to wrap the EM to implement the resetting (as the Doctrine EM becomes unusable once closed and cannot be reset cleanly), the service creating using `$em->getMetadataFactory()` to inject it here would keep referencing the actual EM used to create it, and so might be the wrong one after resetting.

So we might want to inject the EntityManager itself, and not keep a direct reference to the ClassMetadataFactory to play well with resetting.